pkg/interceptor/pacerinteceptor.go

```diff
@@ -0,0 +1,108 @@
+package interceptor
+
+import (
+	"sync"
+
+	"github.com/livekit/mediatransportutil/pkg/pacer"
+	"github.com/pion/interceptor"
+	"github.com/pion/rtp"
+)
+
+type PacerInterceptorFactory struct {
+	pacer pacer.Factory
+	pool  *PacketPool
+}
+
+func NewPacerInterceptorFactory(pacer pacer.Factory) *PacerInterceptorFactory {
+	return &PacerInterceptorFactory{
+		pacer: pacer,
+		pool:  NewPacketPool(500, 1500),
+	}
+}
+
+func (p *PacerInterceptorFactory) NewInterceptor(id string) (interceptor.Interceptor, error) {
+	pacer, err := p.pacer.NewPacer()
+	if err != nil {
+		return nil, err
+	}
+	return &PacerInterceptor{
+		pacer: pacer,
+		pool:  p.pool,
+	}, nil
+}
+
+// PacerInterceptor is an interceptor that paces outgoing packets to avoid congestion on the bursty traffic of huge frames.
+type PacerInterceptor struct {
+	interceptor.NoOp
+
+	pool  *PacketPool
+	pacer pacer.Pacer
+}
+
+// BindRTCPWriter lets you modify any outgoing RTCP packets. It is called once per PeerConnection. The returned method
+// will be called once per packet batch.
+func (pi *PacerInterceptor) BindRTCPWriter(writer interceptor.RTCPWriter) interceptor.RTCPWriter {
+	pi.pacer.Start()
+	return writer
+}
+
+func (pi *PacerInterceptor) BindLocalStream(stream *interceptor.StreamInfo, writer interceptor.RTPWriter) interceptor.RTPWriter {
+	var (
+		lock         sync.Mutex
+		lastSeq      uint16
+		firstPktSeen bool
+	)
+	pacerWriter := func(header *rtp.Header, payload []byte) (int, error) {
+		return writer.Write(header, payload, interceptor.Attributes{})
+	}
+	return interceptor.RTPWriterFunc(func(header *rtp.Header, payload []byte, attributes interceptor.Attributes) (int, error) {
+		// send packets immediately if it's retransmit to reduce frame freeze
+		var isRetransmit bool
+		seq := header.SequenceNumber
+		lock.Lock()
+		if firstPktSeen {
+			if diff := seq - lastSeq; diff == 0 || diff > 0x8000 {
+				isRetransmit = true
+			} else {
+				lastSeq = seq
+			}
+		} else {
+			firstPktSeen = true
+			lastSeq = seq
+		}
+		lock.Unlock()
+		if isRetransmit {
+			return writer.Write(header, payload, attributes)
+		}
+
+		pktSize := len(payload) + header.MarshalSize()
+		poolEntity, pool := pi.pool.Get(pktSize)
+		buf := *poolEntity
+		n, err := header.MarshalTo(buf)
+		if err != nil {
+			if pool != nil {
+				pool.Put(poolEntity)
+			}
+			return 0, err
+		}
+		copy(buf[n:], payload)
+		var headCopy rtp.Header
+		headCopy.Unmarshal(buf)
```

Comment:
Do we need to put together the full packet and unmarshal again to get the header copy? 

In SFU case, the `payload` is in the pool. Here, it is passing along the incoming payload to the pacer, but also creates a pool entity & copies to it. Wondering if that is necessary.

Oh, a copy is needed as pacer could send it later? In that case, I am guessing the `Payload` field 👇 should actuallly point to the `payload` part of the pool entity? Also, wondering if the header can be copied with actually copying it into the pool buffer and unmarshalling it?

Reply:
> In that case, I am guessing the Payload field 👇 should actuallly point to the payload part of the pool entity
You are right, I was missing it when clean the code, thanks!

> Also, wondering if the header can be copied with actually copying it into the pool buffer and unmarshalling it?
We can't do that because the extensions of the header might point to the underlying buffer that could be reused after Enqueue